File: src/endstone_endweave/protocol/v898_to_v924/protocol.py
"""Protocol factory for v898 (1.21.130) server <- v924 (1.26.0) client."""

from endstone_endweave.protocol import Protocol
from endstone_endweave.protocol.mappings.v898_v924 import MAPPINGS
from endstone_endweave.protocol.packet_ids import PacketId
from endstone_endweave.protocol.v898_to_v924.handlers.biome_definition_list import rewrite_biome_definition_list
from endstone_endweave.protocol.v898_to_v924.handlers.book_edit import (
    rewrite_book_edit,
)
from endstone_endweave.protocol.v898_to_v924.handlers.camera import (
    rewrite_camera_instruction,
)
from endstone_endweave.protocol.v898_to_v924.handlers.camera_aim_assist import (
    rewrite_camera_aim_assist_presets,
)
from endstone_endweave.protocol.v898_to_v924.handlers.data_store import (
    rewrite_clientbound_data_store,
    rewrite_serverbound_data_store,
)
from endstone_endweave.protocol.v898_to_v924.handlers.debug_drawer import (
    rewrite_debug_drawer,
)
from endstone_endweave.protocol.v898_to_v924.handlers.diagnostics import (
    rewrite_diagnostics,
)
from endstone_endweave.protocol.v898_to_v924.handlers.graphics_parameter_override import (
    rewrite_graphics_parameter_override,
)
from endstone_endweave.protocol.v898_to_v924.handlers.start_game import (
    rewrite_start_game,
)
from endstone_endweave.protocol.v924_to_v898.handlers.text import (
    rewrite_text_clientbound,
    rewrite_text_serverbound,
)
from endstone_endweave.rewriter import SoundRewriter

SERVER_PROTOCOL = 898
CLIENT_PROTOCOL = 924

def create_protocol() -> Protocol:
    """Create a protocol for v898 server <- v924 client."""
    protocol = Protocol(server_protocol=SERVER_PROTOCOL, client_protocol=CLIENT_PROTOCOL)

    # Serverbound rewriters
    protocol.register_serverbound(PacketId.TEXT, rewrite_text_clientbound)
    protocol.register_serverbound(PacketId.SERVERBOUND_DATA_STORE, rewrite_serverbound_data_store)
    protocol.register_serverbound(PacketId.BOOK_EDIT, rewrite_book_edit)
    protocol.register_serverbound(PacketId.SERVERBOUND_DIAGNOSTICS, rewrite_diagnostics)

    # Clientbound rewriters -- v898/v924 format differences
    protocol.register_clientbound(PacketId.START_GAME, rewrite_start_game)
    protocol.register_clientbound(PacketId.TEXT, rewrite_text_serverbound)
    protocol.register_clientbound(PacketId.CLIENTBOUND_DATA_STORE, rewrite_clientbound_data_store)
    protocol.register_clientbound(PacketId.CAMERA_AIM_ASSIST_PRESETS, rewrite_camera_aim_assist_presets)
    protocol.register_clientbound(PacketId.GRAPHICS_PARAMETER_OVERRIDE, rewrite_graphics_parameter_override)
    protocol.register_clientbound(PacketId.CAMERA_INSTRUCTION, rewrite_camera_instruction)
    protocol.register_clientbound(PacketId.BIOME_DEFINITION_LIST, rewrite_biome_definition_list)
    protocol.register_clientbound(PacketId.SERVER_SCRIPT_DEBUG_DRAWER, rewrite_debug_drawer)

    # Clientbound rewriters -- LevelSoundEvent remapping
    sound = SoundRewriter(
        sound_remap=MAPPINGS.sound.shift_up,
        actor_data_int_remappers={MAPPINGS.actor_data_sound_key: MAPPINGS.sound.shift_up},
    )
    sound.register(protocol)

    return protocol

File: src/protocol/v898_to_v924/protocol.cpp
/// v898 -> v924 protocol factory.

#include "endweave/protocol/v898_to_v924/protocol.h"

#include "endweave/codec/types/enums.h"
#include "endweave/protocol/packet_ids.h"

namespace endweave::v898_to_v924 {

namespace {

int remap_sound(int v) {
    // Remap LevelSoundEvent from v898 -> v924 (shift >= v898_start by +19)
    if (v >= static_cast<int>(sound_boundaries::v898_start)) // 578
        return v + 19; // v944_start is 597, but v924 undefined is also 597
    return v;
}

} // namespace

ProtocolState create_protocol() {
    auto id = [](PacketId p) { return static_cast<int>(p); };

    auto p = std::make_unique<Protocol>(SERVER_PROTOCOL, CLIENT_PROTOCOL, "v898_to_v924");

    // TODO: implement handlers
    auto stub = [](PacketWrapper& w) { w.passthrough_all(); };

    // Cancel v924 serverbound packets unknown to v898
    p->cancel_serverbound({
        id(PacketId::SERVERBOUND_DATA_STORE), // 332
    });

    // Clientbound handlers
    p->register_clientbound(id(PacketId::START_GAME), stub);                    // 11
    p->register_clientbound(id(PacketId::TEXT), stub);                           // 9
    p->register_clientbound(id(PacketId::CLIENTBOUND_DATA_STORE), stub);        // 330
    p->register_clientbound(id(PacketId::CAMERA_AIM_ASSIST_PRESETS), stub);     // 320
    p->register_clientbound(id(PacketId::GRAPHICS_PARAMETER_OVERRIDE), stub);  // 331
    p->register_clientbound(id(PacketId::CAMERA_INSTRUCTION), stub);           // 300
    p->register_clientbound(id(PacketId::BIOME_DEFINITION_LIST), stub);        // 122
    p->register_clientbound(id(PacketId::SERVER_SCRIPT_DEBUG_DRAWER), stub);   // 328

    // Serverbound handlers
    p->register_serverbound(id(PacketId::TEXT), stub);                    // 9
    p->register_serverbound(id(PacketId::BOOK_EDIT), stub);              // 97
    p->register_serverbound(id(PacketId::SERVERBOUND_DIAGNOSTICS), stub); // 315

    // SoundRewriter
    SoundRewriter sound(remap_sound,
        {{static_cast<int>(ActorDataIDs::HEARTBEAT_SOUND_EVENT), remap_sound}},
        {
            static_cast<int>(ActorDataIDs::AIM_ASSIST_PRIORITY_PRESET_ID),   // 136
            static_cast<int>(ActorDataIDs::AIM_ASSIST_PRIORITY_CATEGORY_ID), // 137
            static_cast<int>(ActorDataIDs::AIM_ASSIST_PRIORITY_ACTOR_ID),    // 138
        });
    sound.register_on(*p);

    return {std::move(p), std::move(sound)};
}

} // namespace endweave::v898_to_v924
